src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -1476,41 +1657,187 @@ namespace Microsoft.FSharp.Core
+                            let isTypeAndGenericArgumentsOK fullname            = isValidGenericType false fullname
+                            let isNotTypeOrIsTypeAndGenericArgumentsOK fullname = isValidGenericType true  fullname
                     
-            /// One of the two unique instances of System.Collections.IEqualityComparer. Implements ER semantics
-            /// where equality on NaN returns "true".
-            let fsEqualityComparerNoHashingER = 
-                { new System.Collections.IEqualityComparer with
-                    override iec.Equals(x:obj,y:obj) = GenericEqualityObj true iec (x,y)  // ER Semantics
-                    override iec.GetHashCode(x:obj) = raise (InvalidOperationException (SR.GetString(SR.notUsedForHashing))) }
+                            // avoid any types that need special handling in GenericEqualityObj
+                            // GenericEqualityObj handles string as a special cases, but internally routes to same equality
+
+                            ty.IsSealed  // covers enum and value types
+                                         // ref types need to be sealed as derived class might implement  IStructuralEquatable
+                            && not (isArray ty)
+                            && (er || (not (ty.Equals typeof<float>)))
+                            && (er || (not (ty.Equals typeof<float32>)))
+                            && isNotTypeOrIsTypeAndGenericArgumentsOK "System.Nullable`1"
+                            && not (isStructuralEquatable ty
+                                    // we accept ValueTuple even though it supports IStructuralEquatable 
+                                    // if all generic arguements pass check
+                                    && not (   isTypeAndGenericArgumentsOK "System.ValueTuple`1"
+                                            || isTypeAndGenericArgumentsOK "System.ValueTuple`2"
+                                            || isTypeAndGenericArgumentsOK "System.ValueTuple`3"
+                                            || isTypeAndGenericArgumentsOK "System.ValueTuple`4"
+                                            || isTypeAndGenericArgumentsOK "System.ValueTuple`5"
+                                            || isTypeAndGenericArgumentsOK "System.ValueTuple`6"
+                                            || isTypeAndGenericArgumentsOK "System.ValueTuple`7"
+                                            || isTypeAndGenericArgumentsOK "System.ValueTuple`8"
+                                            || isTypeAndGenericArgumentsOK "Microsoft.FSharp.Collections.FSharpList`1"
+                                            || isTypeAndGenericArgumentsOK "Microsoft.FSharp.Core.FSharpOption`1"
+                                            || isTypeAndGenericArgumentsOK "Microsoft.FSharp.Core.FSharpValueOption`1"
+                                            || isTypeAndGenericArgumentsOK "Microsoft.FSharp.Core.FSharpResult`2"
```

Comment:
This whole condition is fully duplicated on line 1152. Maybe extract it into a function?

Reply:
not the same - strings

Reply:
I wouldn't bother too much checking this yet. I'm still a while off completing it...

Reply:
this is done in the final PR #5307 btw, so assuming the whole chain of PRs is accepted...